gobblin-metrics/src/main/java/gobblin/metrics/reporter/util/EventSubmitter.java

```diff
@@ -0,0 +1,110 @@
+
+import java.util.Collections;
+import java.util.Map;
+
+import com.google.common.base.Optional;
+import com.google.common.collect.Maps;
+
+import gobblin.metrics.GobblinTrackingEvent;
+import gobblin.metrics.MetricContext;
+
+
+/**
+ * Wrapper around Avro {@link gobblin.metrics.GobblinTrackingEvent.Builder} simplifying handling {@link gobblin.metrics.GobblinTrackingEvent}s.
+ *
+ * <p>
+ *   Instances of this class are immutable. Calling set* methods returns a copy of the calling instance.
+ * </p>
+ */
+public class EventSubmitter {
+
+  private final Map<String, String> metadata;
+  private final String namespace;
+  private final Optional<MetricContext> metricContext;
+
+  public static class Builder {
+    private final Optional<MetricContext> metricContext;
+    private final Map<String, String> metadata;
+    private final String namespace;
+
+    public Builder(MetricContext metricContext, String namespace) {
+      this(Optional.of(metricContext), namespace);
+    }
+
+    public Builder(Optional<MetricContext> metricContext, String namespace) {
+      this.metricContext = metricContext;
+      this.namespace = namespace;
+      this.metadata = Maps.newHashMap();
+    }
+
+    public Builder addMetadata(String key, String value) {
+      this.metadata.put(key, value);
+      return this;
+    }
+
+    public EventSubmitter build() {
+      return new EventSubmitter(this);
+    }
+  }
+
+  private EventSubmitter(Builder builder) {
+    this.metadata = builder.metadata;
+    this.namespace = builder.namespace;
+    this.metricContext = builder.metricContext;
+  }
+
+  /**
+   * Submits the {@link gobblin.metrics.GobblinTrackingEvent} to the {@link gobblin.metrics.MetricContext}.
+   * @param name Name of the event.
+   */
+  public void submit(String name) {
+    submit(name, Collections.<String, String>emptyMap());
+  }
+
+  /**
+   * Submits the {@link gobblin.metrics.GobblinTrackingEvent} to the {@link gobblin.metrics.MetricContext}.
+   * @param name Name of the event.
+   * @param metadataEls List of keys and values for metadata of the form key1, value2, key2, value2, ...
+   */
+  public void submit(String name, String... metadataEls) {
+    if(metadataEls.length % 2 != 0) {
+      throw new IllegalArgumentException("Unmatched keys in metadata elements.");
+    }
+
+    Map<String, String> metadata = Maps.newHashMap();
+    for(int i = 0; i < metadataEls.length/2; i++) {
+      metadata.put(metadataEls[2 * i], metadataEls[2 * i + 1]);
+    }
+    submit(name, metadata);
+  }
+
+  /**
+   * Submits the {@link gobblin.metrics.GobblinTrackingEvent} to the {@link gobblin.metrics.MetricContext}.
+   * @param name Name of the event.
+   * @param additionalMetadata Additional metadata to be added to the event.
+   */
+  public void submit(String name, Map<String, String> additionalMetadata) {
+    if(this.metricContext.isPresent()) {
+      Map<String, String> finalMetadata = Maps.newHashMap(this.metadata);
```

Comment:
You can use a `ImmutableMap.Builder` to build the map, the `Builder` is mutable.